File: live/broker.py
"""Order execution layer.

PaperBroker  (default) — matches the strategy's assumed fills (entry at the PoC
              level, stops/TP at levels, SL first), compounds a paper balance,
              subtracts cfg.paper_commission_pct (round-trip % price) per close.

TestnetBroker / LiveBroker — mirror strategy events to Binance
              (spot: testnet.binance.vision / api.binance.com — LIMIT entries,
              stop-limit SL + LIMIT TP pair;
              futures: testnet.binancefuture.com / fapi.binance.com — LIMIT
              entries, reduceOnly STOP_MARKET SL + TAKE_PROFIT_MARKET TP),
              re-armed whenever the trail moves the stop. NOTE: some sandboxes
              are geo-blocked from Binance (HTTP 451) so these are shipped as
              reference code — always validate on testnet from your own machine
              before enabling live.

Every order method is a no-op unless cfg.tradable == True (requires
MODE=testnet|live + keys + BINANCE_DRY_RUN=0). Dry-run is always safe.
"""
import time, hmac, hashlib, json
import urllib.request, urllib.parse


class PaperBroker:
    def __init__(self, cfg):
        self.cfg = cfg
        self.start_equity = cfg.paper_start_usdt
        self.equity = cfg.paper_start_usdt
        self.trades = []
        self.equity_curve = [dict(ts=time.time(), eq=cfg.paper_start_usdt)]

    def on_open(self, ev):
        return None

    def on_trail(self, ev):
        return None

    def apply_trade(self, ev):
        pct = ev["pct"]
        self.equity = self.equity * (1 + (pct - self.cfg.paper_commission_pct) / 100.0)
        self.trades.append(ev)
        self.equity_curve.append(dict(ts=ev.get("exit_ts") or time.time(), eq=self.equity))

    def on_close(self, ev):
        self.apply_trade(ev)

    def balance(self):
        return dict(broker="paper", equity=self.equity, start=self.start_equity,
                    n_trades=len(self.trades), unrealized=None)


class _BinanceRest:
    qty_prec = 6          # quantity rounding (spot default; futures overrides)

    def __init__(self, key, secret, base):
        self.key, self.secret, self.base = key, secret, base

    def _call(self, method, path, params=None, signed=False):
        url = self.base + path
        headers = {"X-MBX-APIKEY": self.key}
        body = None
        if signed:
            params = dict(params or {})
            params["timestamp"] = int(time.time() * 1000)
            params["recvWindow"] = 5000
            params["signature"] = hmac.new(self.secret.encode(),
                                           urllib.parse.urlencode(params).encode(),
                                           hashlib.sha256).hexdigest()
            body = urllib.parse.urlencode(params).encode()
        elif params:
            url += "?" + urllib.parse.urlencode(params)
        req = urllib.request.Request(url, data=body, headers=headers, method=method)
        with urllib.request.urlopen(req, timeout=15) as r:
            return json.loads(r.read().decode())

    def ping(self):
        return self._call("GET", "/api/v3/ping")

    def account(self):
        return self._call("GET", "/api/v3/account", signed=True)

    def place_limit(self, symbol, side, qty, price):
        return self._call("POST", "/api/v3/order", signed=True, params=dict(
            symbol=symbol, side=side, type="LIMIT", timeInForce="GTC",
            quantity=f"{qty:.8f}".rstrip("0").rstrip("."),
            price=f"{price:.2f}"))

    def place_stop_limit(self, symbol, side, qty, stop_price, limit_price):
        return self._call("POST", "/api/v3/order", signed=True, params=dict(
            symbol=symbol, side=side, type="STOP_LOSS_LIMIT", timeInForce="GTC",
            quantity=f"{qty:.8f}".rstrip("0").rstrip("."),
            stopPrice=f"{stop_price:.2f}", price=f"{limit_price:.2f}"))

    def place_take_profit(self, symbol, side, qty, price):
        # spot: plain GTC limit at the TP level
        return self.place_limit(symbol, side, qty, price)

    def cancel_all(self, symbol):
        return self._call("DELETE", "/api/v3/openOrders", signed=True,
                          params={"symbol": symbol})


class _BinanceFuturesRest(_BinanceRest):
    """USDT-M perpetual (fapi) order routing: LIMIT entries, reduceOnly
    STOP_MARKET SL / TAKE_PROFIT_MARKET TP, cancel-all via allOpenOrders.
    Lot precision / min notional are read from the venue's exchangeInfo."""

    def __init__(self, key, secret, base, symbol=""):
        super().__init__(key, secret, base)
        self.qty_prec = 3
        self.min_notional = 5.0
        if symbol:
            try:
                info = self._call("GET", "/fapi/v1/exchangeInfo")
                for s in info.get("symbols", []):
                    if s.get("symbol") == symbol.upper():
                        self.qty_prec = int(s.get("quantityPrecision", 3))
                        for f in s.get("filters", []):
                            if f.get("filterType") == "MIN_NOTIONAL":
                                self.min_notional = float(f.get("notional", 5))
                        break
            except Exception:
                pass   # keep defaults; orders will surface any mismatch

    def _qty(self, q):
        return f"{q:.{self.qty_prec}f}"

    def place_limit(self, symbol, side, qty, price):
        return self._call("POST", "/fapi/v1/order", signed=True, params=dict(
            symbol=symbol, side=side, type="LIMIT", timeInForce="GTC",
            quantity=self._qty(qty), price=f"{price:.2f}"))

    def place_stop_limit(self, symbol, side, qty, stop_price, limit_price):
        # futures: protective stop as a reduce-only STOP_MARKET at the level
        return self._call("POST", "/fapi/v1/order", signed=True, params=dict(
            symbol=symbol, side=side, type="STOP_MARKET", reduceOnly="true",
            quantity=self._qty(qty), stopPrice=f"{stop_price:.2f}"))

    def place_take_profit(self, symbol, side, qty, price):
        # futures: TP as a reduce-only TAKE_PROFIT_MARKET at the level
        return self._call("POST", "/fapi/v1/order", signed=True, params=dict(
            symbol=symbol, side=side, type="TAKE_PROFIT_MARKET",
            reduceOnly="true", quantity=self._qty(qty),
            stopPrice=f"{price:.2f}"))

    def cancel_all(self, symbol):
        return self._call("DELETE", "/fapi/v1/allOpenOrders", signed=True,
                          params={"symbol": symbol})

    def set_leverage(self, symbol, leverage=1):
        return self._call("POST", "/fapi/v1/leverage", signed=True,
                          params=dict(symbol=symbol, leverage=leverage))

    def balance(self):
        return self._call("GET", "/fapi/v2/balance", signed=True)


class BinanceBroker:
    """Mirrors strategy events onto Binance (spot or USDT-M futures per cfg).
    Falls back to a logged status on any error and always keeps the paper
    record as the PnL source of truth."""

    def __init__(self, cfg, base_url, venue="spot"):
        self.cfg = cfg
        self.venue = venue
        if venue == "futures":
            self.cli = _BinanceFuturesRest(cfg.api_key, cfg.api_secret,
                                           base_url, cfg.symbol)
        else:
            self.cli = _BinanceRest(cfg.api_key, cfg.api_secret, base_url)
        self.status = []
        self._mirror = None          # dict(side, qty, sl, tp, armed)
        self.paper = PaperBroker(cfg)
        if venue == "futures" and cfg.tradable:
            try:                     # keep the mirror 1x (no leverage surprises)
                self.cli.set_leverage(cfg.symbol, 1)
                self._log(f"leverage set to 1x on {cfg.symbol}")
            except Exception as ex:
                self._log(f"leverage set failed: {ex}")

    def _log(self, msg):
        d = dict(ts=time.time(), msg=str(msg))
        self.status.append(d)
        if len(self.status) > 200:
            self.status = self.status[-150:]
        return d

    def _tradable(self):
        return self.cfg.tradable

    def _arm(self):
        m = self._mirror
        if not (self._tradable() and m):
            return
        try:
            self.cli.cancel_all(self.cfg.symbol)
            close_side = "SELL" if m["side"] == "long" else "BUY"
            if m.get("sl"):
                self.cli.place_stop_limit(self.cfg.symbol, close_side, m["qty"],
                                          m["sl"], m["sl"])
            if m.get("tp"):
                self.cli.place_take_profit(self.cfg.symbol, close_side,
                                           m["qty"], m["tp"])
            m["armed"] = True
            self._log(f"protection armed: SL {m.get('sl')} TP {m.get('tp')}")
        except Exception as ex:
            m["armed"] = False
            self._log(f"arm failed: {ex}")

    def on_open(self, ev):
        q = round(self.cfg.position_usd / ev["entry"], self.cli.qty_prec)
        if self._tradable():
            min_not = getattr(self.cli, "min_notional", None)
            if min_not and q * ev["entry"] < min_not:
                self._log(f"entry skipped: qty {q} x {ev['entry']:.2f} below "
                          f"venue min notional {min_not}")
            else:
                try:
                    order = self.cli.place_limit(self.cfg.symbol,
                                                 "BUY" if ev["side"] == "long" else "SELL",
                                                 q, ev["entry"])
                    self._log(f"entry {order.get('orderId')} {ev['side']} {q} @ {ev['entry']:.2f}")
                except Exception as ex:
                    self._log(f"entry order failed: {ex}")
        else:
            self._log(f"(dry) entry {ev['side']} {q} @ {ev['entry']:.2f}")
        self._mirror = dict(side=ev["side"], qty=q, sl=ev["sl"], tp=ev["tp"], armed=False)
        if self._tradable():
            self._arm()

    def on_trail(self, ev):
        if self._mirror:
            self._mirror["sl"] = ev["sl"]
        if self._tradable():
            self._arm()
        else:
            self._log(f"(dry) trail -> {ev['a']}R SL {ev['sl']:.2f}")

    def on_close(self, ev):
        self.paper.on_close(ev)
        if self._tradable():
            try:
                self.cli.cancel_all(self.cfg.symbol)
                self._log(f"closed; orders cancelled (equity {self.paper.equity:,.2f})")
            except Exception as ex:
                self._log(f"cancel on close failed: {ex}")
        else:
            self._log(f"closed {ev['side']} {ev['exit_type']} @ {ev['exit']:.2f} "
                      f"({ev['pct']:+.3f}%) equity {self.paper.equity:,.2f}")
        self._mirror = None

    def balance(self):
        b = self.paper.balance()
        b["broker"] = self.cfg.resolved_mode
        b["mirror_armed"] = bool(self._mirror and self._mirror.get("armed"))
        return b


def make_broker(cfg):
    mode = cfg.resolved_mode
    if cfg.venue == "futures":
        if mode == "live":
            return BinanceBroker(cfg, "https://fapi.binance.com", venue="futures")
        if mode == "testnet":
            return BinanceBroker(cfg, "https://testnet.binancefuture.com",
                                 venue="futures")
        return PaperBroker(cfg)
    if mode == "live":
        return BinanceBroker(cfg, "https://api.binance.com")
    if mode == "testnet":
        return BinanceBroker(cfg, "https://testnet.binance.vision")
    return PaperBroker(cfg)

File: rust/src/broker.rs
//! Order execution layer — mirrors strategy events to Binance (futures
//! testnet/mainnet: LIMIT entries, reduceOnly STOP_MARKET SL +
//! TAKE_PROFIT_MARKET TP, re-armed on trail moves; spot: LIMIT entry,
//! stop-limit + limit pair). The paper Ledger remains the PnL source of truth.
//!
//! Every order method is a no-op unless cfg.tradable() (MODE=testnet|live +
//! keys + BINANCE_DRY_RUN=0). Dry-run is always safe.
//!
//! v2: order quantity comes from the Ledger's sizing (same as paper), fixing
//! the legacy 100x ledger-vs-mirror inconsistency (AUDIT.md fix #3).

use serde_json::{json, Value};
use std::time::Duration;
use crate::config::Config;
use crate::engine::Trade;
use crate::ledger::{FeeModel, Ledger, Sizing, SizingCfg};

pub struct RestCli {
    key: String,
    secret: String,
    base: String,
    pub qty_prec: u32,
    pub min_notional: f64,
}

impl RestCli {
    pub fn new(key: &str, secret: &str, base: &str, symbol: &str, futures: bool) -> RestCli {
        let mut cli = RestCli {
            key: key.to_string(),
            secret: secret.to_string(),
            base: base.to_string(),
            qty_prec: if futures { 3 } else { 6 },
            min_notional: 5.0,
        };
        if futures && !symbol.is_empty() {
            // read lot precision / min notional from exchangeInfo (best effort)
            if let Ok(info) = cli.call("GET", "/fapi/v1/exchangeInfo", &[], false) {
                if let Some(syms) = info.get("symbols").and_then(|v| v.as_array()) {
                    for s in syms {
                        if s.get("symbol").and_then(|v| v.as_str()) == Some(symbol) {
                            cli.qty_prec = s
                                .get("quantityPrecision")
                                .and_then(|v| v.as_u64())
                                .unwrap_or(3) as u32;
                            if let Some(filters) = s.get("filters").and_then(|v| v.as_array()) {
                                for f in filters {
                                    if f.get("filterType").and_then(|v| v.as_str()) == Some("MIN_NOTIONAL") {
                                        cli.min_notional = f
                                            .get("notional")
                                            .and_then(|v| v.as_f64())
                                            .unwrap_or(5.0);
                                    }
                                }
                            }
                            break;
                        }
                    }
                }
            }
        }
        cli
    }

    fn call(&self, method: &str, path: &str, params: &[(&str, String)], signed: bool) -> Result<Value, String> {
        let mut p: Vec<(String, String)> = params
            .iter()
            .map(|(k, v)| (k.to_string(), v.clone()))
            .collect();
        let mut url = format!("{}{}", self.base, path);
        let body: Option<String>;
        if signed {
            p.push(("timestamp".into(), format!("{}", now_ms())));
            p.push(("recvWindow".into(), "5000".into()));
            let qs = urlencode(&p);
            let sig = hmac_sha256_hex(self.secret.as_bytes(), qs.as_bytes());
            body = Some(format!("{}&signature={}", qs, sig));
        } else {
            body = None;
            if !p.is_empty() {
                url.push('?');
                url.push_str(&urlencode(&p));
            }
        }
        let mut req = ureq::request(method, &url)
            .set("X-MBX-APIKEY", &self.key)
            .timeout(Duration::from_secs(15));
        if body.is_some() {
            req = req.set("Content-Type", "application/x-www-form-urlencoded");
        }
        let resp = match &body {
            Some(b) => req.send_string(b),
            None => req.call(),
        };
        match resp {
            Ok(r) => r.into_json::<Value>().map_err(|e| e.to_string()),
            Err(ureq::Error::Status(code, r)) => {
                let b = r.into_string().unwrap_or_default();
                Err(format!("HTTP {code}: {}", b.chars().take(200).collect::<String>()))
            }
            Err(e) => Err(e.to_string()),
        }
    }

    fn fmt_qty(&self, q: f64) -> String {
        format!("{:.*}", self.qty_prec as usize, q)
    }

    // ---- futures (USDT-M) order routing -----------------------------------
    pub fn fut_limit(&self, symbol: &str, side: &str, qty: f64, price: f64) -> Result<Value, String> {
        self.call("POST", "/fapi/v1/order", &[
            ("symbol", symbol.into()),
            ("side", side.into()),
            ("type", "LIMIT".into()),
            ("timeInForce", "GTC".into()),
            ("quantity", self.fmt_qty(qty)),
            ("price", format!("{price:.2}")),
        ], true)
    }

    pub fn fut_stop(&self, symbol: &str, side: &str, qty: f64, stop: f64) -> Result<Value, String> {
        self.call("POST", "/fapi/v1/order", &[
            ("symbol", symbol.into()),
            ("side", side.into()),
            ("type", "STOP_MARKET".into()),
            ("reduceOnly", "true".into()),
            ("quantity", self.fmt_qty(qty)),
            ("stopPrice", format!("{stop:.2}")),
        ], true)
    }

    pub fn fut_tp(&self, symbol: &str, side: &str, qty: f64, stop: f64) -> Result<Value, String> {
        self.call("POST", "/fapi/v1/order", &[
            ("symbol", symbol.into()),
            ("side", side.into()),
            ("type", "TAKE_PROFIT_MARKET".into()),
            ("reduceOnly", "true".into()),
            ("quantity", self.fmt_qty(qty)),
            ("stopPrice", format!("{stop:.2}")),
        ], true)
    }

    pub fn fut_cancel_all(&self, symbol: &str) -> Result<Value, String> {
        self.call("DELETE", "/fapi/v1/allOpenOrders", &[("symbol", symbol.into())], true)
    }

    pub fn fut_set_leverage(&self, symbol: &str, leverage: u32) -> Result<Value, String> {
        self.call("POST", "/fapi/v1/leverage", &[
            ("symbol", symbol.into()),
            ("leverage", leverage.to_string()),
        ], true)
    }

    // ---- spot order routing -----------------------------------------------
    pub fn spot_limit(&self, symbol: &str, side: &str, qty: f64, price: f64) -> Result<Value, String> {
        self.call("POST", "/api/v3/order", &[
            ("symbol", symbol.into()),
            ("side", side.into()),
            ("type", "LIMIT".into()),
            ("timeInForce", "GTC".into()),
            ("quantity", self.fmt_qty(qty)),
            ("price", format!("{price:.2}")),
        ], true)
    }

    pub fn spot_stop_limit(&self, symbol: &str, side: &str, qty: f64, stop: f64, limit: f64) -> Result<Value, String> {
        self.call("POST", "/api/v3/order", &[
            ("symbol", symbol.into()),
            ("side", side.into()),
            ("type", "STOP_LOSS_LIMIT".into()),
            ("timeInForce", "GTC".into()),
            ("quantity", self.fmt_qty(qty)),
            ("stopPrice", format!("{stop:.2}")),
            ("price", format!("{limit:.2}")),
        ], true)
    }

    pub fn spot_cancel_all(&self, symbol: &str) -> Result<Value, String> {
        self.call("DELETE", "/api/v3/openOrders", &[("symbol", symbol.into())], true)
    }
}

fn now_ms() -> u128 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis()
}

fn escape(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.bytes() {
        match c {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => out.push(c as char),
            _ => out.push_str(&format!("%{:02X}", c)),
        }
    }
    out
}

fn urlencode(params: &[(String, String)]) -> String {
    params
        .iter()
        .map(|(k, v)| format!("{}={}", escape(k), escape(v)))
        .collect::<Vec<_>>()
        .join("&")
}

fn hmac_sha256_hex(key: &[u8], msg: &[u8]) -> String {
    use ring::hmac::{Key, HMAC_SHA256};
    let key = Key::new(HMAC_SHA256, key);
    let tag = ring::hmac::sign(&key, msg);
    tag.as_ref().iter().map(|b| format!("{:02x}", b)).collect()
}

// ---------------------------------------------------------------------------

struct MirrorState {
    side: i8,
    qty: f64,
    sl: f64,
    tp: f64,
    armed: bool,
}

pub struct Broker {
    cfg: Config,
    cli: Option<RestCli>,
    pub ledger: Ledger,
    status: Vec<Value>,
    mirror: Option<MirrorState>,
    sizing: Option<Sizing>,
}

impl Broker {
    pub fn new(cfg: &Config) -> Broker {
        let fee = FeeModel {
            maker_pct: cfg.fee_maker_pct,
            taker_pct: cfg.fee_taker_pct,
            slip_pct: cfg.sl_slip_pct,
        };
        let scfg = SizingCfg {
            legacy: cfg.legacy_sizing,
            position_usd: cfg.position_usd,
            risk_pct: cfg.risk_pct,
            leverage_cap: cfg.leverage_cap,
        };
        let ledger = Ledger::new(cfg.paper_start_usdt, scfg, fee);
        let mut b = Broker {
            cfg: cfg.clone(),
            cli: None,
            ledger,
            status: Vec::new(),
            mirror: None,
            sizing: None,
        };
        if cfg.tradable() && !cfg.api_key.is_empty() {
            let futures = cfg.venue == "futures";
            let cli = RestCli::new(&cfg.api_key, &cfg.api_secret, &cfg.order_base(), &cfg.symbol, futures);
            if futures {
                match cli.fut_set_leverage(&cfg.symbol, 1) {
                    Ok(_) => b.log(format!("leverage set to 1x on {}", cfg.symbol)),
                    Err(e) => b.log(format!("leverage set failed: {e}")),
                }
            }
            b.cli = Some(cli);
        }
        b
    }

    fn log(&mut self, msg: String) {
        let d = json!({"ts": now_ms() as f64 / 1000.0, "msg": msg});
        self.status.push(d);
        if self.status.len() > 200 {
            let drop = self.status.len() - 150;
            self.status.drain(0..drop);
        }
        println!("[broker] {msg}");
    }

    pub fn status_json(&self) -> Value {
        self.status.iter().rev().take(20).cloned().collect::<Vec<_>>().into()
    }

    fn tradable(&self) -> bool {
        self.cfg.tradable() && self.cli.is_some()
    }

    fn arm(&mut self) {
        let (side, qty, sl, tp) = match &self.mirror {
            Some(m) => (m.side, m.qty, m.sl, m.tp),
            None => return,
        };
        if !self.tradable() {
            return;
        }
        let close_side = if side == 1 { "SELL" } else { "BUY" };
        let futures = self.cfg.venue == "futures";
        let symbol = self.cfg.symbol.clone();
        let res = match &self.cli {
            None => return,
            Some(cli) => (|| -> Result<(), String> {
                if futures {
                    cli.fut_cancel_all(&symbol)?;
                    cli.fut_stop(&symbol, close_side, qty, sl)?;
                    cli.fut_tp(&symbol, close_side, qty, tp)?;
                } else {
                    cli.spot_cancel_all(&symbol)?;
                    cli.spot_stop_limit(&symbol, close_side, qty, sl, sl)?;
                    cli.spot_limit(&symbol, close_side, qty, tp)?;
                }
                Ok(())
            })(),
        };
        match res {
            Ok(()) => {
                if let Some(m) = &mut self.mirror {
                    m.armed = true;
                }
                self.log(format!("protection armed: SL {sl} TP {tp}"));
            }
            Err(e) => {
                if let Some(m) = &mut self.mirror {
                    m.armed = false;
                }
                self.log(format!("arm failed: {e}"));
            }
        }
    }

    /// Live open event: size from the ledger, mirror the entry if tradable.
    pub fn on_open(&mut self, side: i8, entry: f64, sl: f64, tp: f64, slp: f64) {
        let sizing = self.ledger.on_open(entry, slp);
        self.sizing = Some(sizing);
        let prec = self.cli.as_ref().map(|c| c.qty_prec).unwrap_or(3);
        let min_notional = self.cli.as_ref().map(|c| c.min_notional).unwrap_or(5.0);
        let q = round_to(sizing.qty, prec);
        let tradable = self.tradable();
        let futures = self.cfg.venue == "futures";
        let symbol = self.cfg.symbol.clone();
        let order_side = if side == 1 { "BUY" } else { "SELL" };
        if tradable {
            if q * entry < min_notional {
                self.log(format!(
                    "entry skipped: qty {q} x {entry:.2} below venue min notional {min_notional}"
                ));
            } else {
                let r = match &self.cli {
                    None => return,
                    Some(cli) => {
                        if futures {
                            cli.fut_limit(&symbol, order_side, q, entry)
                        } else {
                            cli.spot_limit(&symbol, order_side, q, entry)
                        }
                    }
                };
                match r {
                    Ok(o) => self.log(format!(
                        "entry order {} {} qty {q} @ {entry:.2} (notional {:.0} USDT, risk {:.2} USDT)",
                        o.get("orderId").map(|v| v.to_string()).unwrap_or_default(),
                        if side == 1 { "long" } else { "short" },
                        sizing.notional_usd, sizing.risk_usd
                    )),
                    Err(e) => self.log(format!("entry order failed: {e}")),
                }
            }
        } else {
            self.log(format!(
                "(dry) entry {} qty {q} @ {entry:.2} (notional {:.0} USDT, risk {:.2} USDT)",
                if side == 1 { "long" } else { "short" },
                sizing.notional_usd, sizing.risk_usd
            ));
        }
        self.mirror = Some(MirrorState { side, qty: q, sl, tp, armed: false });
        self.arm();
    }

    pub fn on_trail(&mut self, sl: f64) {
        if let Some(m) = &mut self.mirror {
            m.sl = sl;
        }
        if self.tradable() {
            self.arm();
        } else {
            self.log(format!("(dry) trail -> SL {sl:.2}"));
        }
    }

    /// Live close event: apply the ledger (fees + sizing) and clean up orders.
    /// Returns the enriched close payload.
    pub fn on_close(&mut self, t: &Trade) -> Value {
        let enriched = self.ledger.on_close(t, t.exit_ts);
        let tradable = self.tradable();
        let futures = self.cfg.venue == "futures";
        let symbol = self.cfg.symbol.clone();
        if tradable {
            let r = match &self.cli {
                None => None,
                Some(cli) => Some(if futures {
                    cli.fut_cancel_all(&symbol)
                } else {
                    cli.spot_cancel_all(&symbol)
                }),
            };
            match r {
                Some(Ok(_)) => self.log(format!(
                    "closed {} @ {:.2} net {} USDT (fees {}) equity {}",
                    t.exit_type, t.exit, enriched["pnl_usd"], enriched["fee_usd"], enriched["equity"]
                )),
                Some(Err(e)) => self.log(format!("cancel on close failed: {e}")),
                None => {}
            }
        } else {
            self.log(format!(
                "closed {} @ {:.2} net {} USDT (fees {}) equity {}",
                t.exit_type, t.exit, enriched["pnl_usd"], enriched["fee_usd"], enriched["equity"]
            ));
        }
        self.mirror = None;
        self.sizing = None;
        enriched
    }

    /// Warm-up replay of historical trades: ledger accounting only (no orders,
    /// no hub publishing) — same as the legacy bootstrap behaviour.
    /// Returns the enriched close payload (for the trades log).
    pub fn quiet_close(&mut self, t: &Trade) -> Value {
        self.ledger.apply_warm(t)
    }

    pub fn balance(&self) -> Value {
        let armed = self.mirror.as_ref().map(|m| m.armed).unwrap_or(false);
        self.ledger.balance(&self.cfg.resolved_mode(), armed)
    }

}

fn round_to(q: f64, prec: u32) -> f64 {
    let m = 10f64.powi(prec as i32);
    (q * m).round() / m
}
